docs: add repository documentation (#1)
eduNEXT/openedx-alexa-adaptor-template: README.md

Reviewer: **Observation:** In the numbered instructional sections, you use " " to called out action navigators. Best practice is use **bold text**.  E.g., Line 192-196

2. Go to **More** → **Skills & Games** → **Your Skills**.
3. Select the skill you want to configure. e.g., **Open edX Assistant**.
4. Select **Settings**.
5. Select **Manage account permissions**.
6. Enable **Email Address** permission.

Line 50: comma not required. ... Open edX and retrieve....
Line 122: replace 'to use' with 'using'
Line 137: i believe it should say "on" the sample skill, not 'in'.
Line 138: remove 'the' before the word 'our'.
Line 153: replace you'll with 'you will' (formality)
Line 200: replace haven't with 'have not' (formality)
Line 208: replace don't with 'do not' (formality)
Line 237; verb is 'shows' - add an 's'
Line 238, add comma after 'known profile,' and ...
Line 250-251: e.g., (add comma)
Line 255: replace can't with 'cannot' (formality)
Line 277 and 280: verb is 'updates' - add an 's'

interactions:

- **Launch the skill**: "Alexa, open educational assistant."

  > _Alexa Response_: "Welcome \<first_name\>, this is the Open edX assistant,
  > I can provide you information about student metrics and important aspects
  > of a course."

  e.g:

  > "_Alexa Response_: Welcome **John**, this is the Open edX assistant, I can
  > provide you information about student metrics and important aspects of a
  > course."

- **Get course progress**: "Alexa, give me my progress in the course of
  \<course-name\>."

  > _Alexa Response_: "The progress for the student with
  > username \<username\> in the course of \<coursename\> is \<percentage\>.

  e.g:

  > "_Alexa Response_: The progress for the student with username **johndoe**
  > in the course of **Introduction to Linux** is **50%**."

## Creating the Skill

### Prerequisites

Before setting up and running this sample skill, ensure you have the following
prerequisites:

1. An [Amazon Developer account](https://developer.amazon.com/) for creating
   and configuring the skill.
2. Installed [Node.js and npm](https://nodejs.org/en), [Python 3.8](https://www.python.org/downloads/),
   and [git](https://git-scm.com/).

> 3. Access to an Open edX platform with admin privileges.

Reviewer 2: Why do I need admin privileges?

Author: In order to create the necessary application in the Django OAuth Toolkit

Reviewer 2: admin refers to django staff or superuser?

Author: I was able to verify that with **staff** permissions, endpoints can be consumed.

Reviewer 2: What about a user without staff/superuser permissions but with eox core API permissions? 
![image](https://github.com/eduNEXT/openedx-alexa-adaptor-template/assets/64440265/59abec4b-b962-4dd4-a552-43469111c47c)

Author: In this case, it would work for the eox-core API, but not for the edX API, to get the enrollments we need to have staff permissions.

Reviewer 2: Good to know! Thanks

4. ASK CLI installed and configured in your computer. Follow the steps in the
   [ASK CLI documentation](./docs/1-ask-cli.md)
5. `eox-core` installed and configured in your Open edX platform. Follow the
   steps in the [eox-core documentation](./docs/2-eox-core.md)

### Development

> We created a `Makefile` to help you to prepare your environment and init with

Reviewer 2: ```suggestion
We created a `Makefile` to help you prepare and initialize your environment for Alexa's skill development.
```

> the development of the skill. Before to use the commands, ensure that you have

Reviewer 2: ```suggestion
the development of the skill. Before using the commands, ensure that you have
```

the following requirements:

- Python 3.8, we recommend to use a virtual environment to avoid conflicts with
  other python projects.
- ASK CLI, follow the steps in the [ASK CLI documentation](./docs/1-ask-cli.md)

After to ensure that you have the requirements, you only need to execute the
following command:

```bash
make bootstrap
```

This command allows you initialize and configure the development environment.
Under the hood, the command executes the following commands:

- `make configure`: Configure the ASK CLI with your Amazon Developer account.
- `make setup`: Create a new skill based in the sample skill. This command
  creates a new skill based in the our sample-skill in the Alexa Developer
  Console, and a new git repository in the `skills/` folder.

### Deployment

> Make deploy is very easy!. As the skill is [alexa-hosted](https://developer.amazon.com/en-US/docs/alexa/hosted-skills/build-a-skill-end-to-end-using-an-alexa-hosted-skill.html),

Reviewer 2: ```suggestion
As the skill is [alexa-hosted](https://developer.amazon.com/en-US/docs/alexa/hosted-skills/build-a-skill-end-to-end-using-an-alexa-hosted-skill.html),
```

> you can deploy the skill using only git commands. Follow the steps below:

Reviewer 2: ```suggestion
you can deploy the skill using only git commands. So, deployment is very straightforward! Follow the steps below:
```

   `<skill-name>` is the name of your skill. e.g: `skills/openedx-skill/`.
2. Make your changes.
3. Commit your changes.
4. Push your changes to the repository. Ensure that you push to the `master` branch.

After to push your changes, the skill will be deployed automatically, only you
need to wait a few minutes to see the changes in the skill in the Alexa
Developer Console.

## Configuring the Skill

### Adding Environment Variables

This skill requires some environment variables to work properly. Follow the
steps below to configure the environment variables:

1. Go to the created skill in the `skills/<skill-name>/` folder, where
   `<skill-name>` is the name of your skill. e.g: `skills/openedx-skill/`.
2. Create a `.env` file in the `lambda/alexa/` folder. e.g: `skills/openedx-skill/lambda/alexa/.env`.
3. Add the following environment variables to the `.env` file:

   ```bash
    SKILL_PROFILE_EMAIL_BACKEND=<your-email-backend> # e.g: auth.backends.custom.CustomEmailBackend
    LMS_DOMAIN=<your-lms-domain> # e.g: https://lms.example.com
    EOX_CORE_CLIENT_ID=<your-eox-core-client-id>
    EOX_CORE_CLIENT_SECRET=<your-eox-core-client-secret>

>     EOX_CORE_GRANT_TYPE=client_credentials

Reviewer 2: Can we explain that we got this from configuring eox-core?

### Configuring Permissions

This skill requires some permissions in the Alexa account to work properly.
Follow the steps below to configure the permissions:

1. Open the Alexa app (Android or iOS). **IMPORTANT**: The account must be the
   same account that you use to create the skill in the Alexa Developer
   Console.
2. Go to "More" → "Skills & Games" → "Your Skills".
3. Select the skill you want to configure. e.g: "Open edX Assistant".
4. Select "Settings".
5. Select "Manage account permissions".
6. Enable "Email Address" permission.
7. Save the changes.

Additionally, if you want to use the skill you need to enable the "Voice
Profile" permission. Follow the steps below to enable (if you don't have
enabled) the "Voice Profile" permission:

1. Open the Alexa app (Android or iOS). **IMPORTANT**: The account must be the
   same account that you use to create the skill in the Alexa Developer
   Console.
2. Go to "More" → "Settings" → "Your Profile & Family".
3. Select your profile.
4. Select "Voice ID".
5. If you don't have set up your Voice ID, perform the setup.
6. Go to "Voice ID settings".
7. Enable "Personalize Skills" permission.

### Enabling Spanish Support

> As default, the skill only is available in English. If you want to activate

Reviewer 2: ```suggestion
By default, the skill only recognizes English. If you want to activate
```

Spanish, you need to follow the steps below:

1. Go to the [Alexa Developer Console](https://developer.amazon.com/alexa/console/ask).
2. Select the skill you want to configure.
3. Select the "Build" tab.
4. Select "Language Settings".
5. Select "+ Add New Language".
6. Select "Spanish (ES)".
7. Select "Save Changes".

## Testing the Skill

### Testing in Amazon Developer Console

To test the sample Alexa skill, follow these steps:

1. Go to the [Amazon Developer Console](https://developer.amazon.com/alexa/console/ask).
2. Select the Skill you want to test.
3. Select the "Test" tab.
4. Select "Development" from the "Skill testing is enabled in:" dropdown.
5. As the Skill uses voice profiles, to test interactions you need to do it by
   voice. If you are known profile, you can interact with the skill without

>    problems. If you are not a known profile, the skill show you a message

Reviewer 2: ```suggestion
   any issues. If you are not a known profile, the skill show you a message
```

   indicating that you are not a known profile and the interaction will finish.

> Also, you can test the skill using the Alexa app, only ensure that you are

Reviewer 2: ```suggestion
Also, you can test the skill using the Alexa app, be sure you are
```

> using the same account that you use to create the skill in the Alexa Developer

Reviewer 2: ```suggestion
using the same account you use to create the skill in the Alexa Developer
```

### Development

We created a `Makefile` to help you to prepare and initialize your environment
for Alexa's skill development. Before using the commands, ensure that you have
the following requirements:

- Python 3.8, we recommend to use a virtual environment to avoid conflicts with
  other python projects.
- ASK CLI, follow the steps in the [ASK CLI documentation](./docs/1-ask-cli.md)

After to ensure that you have the requirements, you only need to execute the
following command:

```bash
make bootstrap
```

This command allows you initialize and configure the development environment.
Under the hood, the command executes the following commands:

- `make configure`: Configure the ASK CLI with your Amazon Developer account.
- `make setup`: Create a new skill based in the sample skill. This command
  creates a new skill based in the our sample-skill in the Amazon Alexa
  Console, and a new git repository in the `skills/` folder.

### Deployment

As the skill is [alexa-hosted](https://developer.amazon.com/en-US/docs/alexa/hosted-skills/build-a-skill-end-to-end-using-an-alexa-hosted-skill.html),
you can deploy the skill using only git commands. So, deployment is very

> straightforward!. Follow the steps below:

Reviewer 2: ```suggestion
straightforward! Follow the steps below:
```

1. Go to the created skill in the `skills/<skill-name>/` folder, where
   `<skill-name>` is the name of your skill. e.g: `skills/openedx-skill/`.
2. Make your changes.
3. Commit your changes.
4. Push your changes to the repository. Ensure that you push to the `master` branch.

> After to push your changes, the skill will be deployed automatically, only you
> need to wait a few minutes to see the changes in the skill in the Amazon Alexa

Reviewer 2: ```suggestion
After you push your changes, the skill will be deployed automatically, you'll only
need to wait a few minutes to see the changes in the skill in the Amazon Alexa
```

    EOX_CORE_CLIENT_ID=<your-eox-core-client-id>
    EOX_CORE_CLIENT_SECRET=<your-eox-core-client-secret>
    EOX_CORE_GRANT_TYPE=client_credentials
    REQUEST_MAX_TIMEOUT=<your-request-max-timeout> # e.g: 5
   ```

   **NOTE**: The `EOX_CORE_CLIENT_ID`, `EOX_CORE_CLIENT_SECRET`, and
   `EOX_CORE_GRANT_TYPE` environment variables are provided by the [`eox-core`
   configuration](./docs/2-eox-core.md) explained in the prerequisites section.

4. Save the changes.
5. Commit and push the changes to the repository.

### Configuring Permissions

This skill requires some permissions in the Alexa account to work properly.
Follow the steps below to configure the permissions:

1. Open the Alexa app (Android or iOS). **IMPORTANT**: The account must be the
   same account that you use to create the skill in the Amazon Alexa Console.
2. Go to "More" → "Skills & Games" → "Your Skills".
3. Select the skill you want to configure. e.g: "Open edX Assistant".
4. Select "Settings".
5. Select "Manage account permissions".
6. Enable "Email Address" permission.
7. Save the changes.

Additionally, if you want to use the skill you need to enable the "Voice

> Profile" permission. Follow the steps below to enable (if you don't have
> enabled) the "Voice Profile" permission:

Reviewer 2: ```suggestion
Profile" permission. Follow the steps below to enable (if you haven't) the "Voice Profile" permission:
```
